### ISSUE TITLE
[bug] [core] some (small?) terrain geotiffs don't load at certain zoom levels

### ISSUE THREAD

**Issue by Mis012** [osmandapp/OsmAnd#21403]
### Description

For some geotiffs (possibly ones covering a relatively small area), the following [1] condition is satisfied for most zoom levels, that is, `tileFound` gets set to `false`. Identical creation options with a larger geotiff (in terms of area on the globe) result in a working geotiff. Notably, commenting this out (leaving `tileFound` as true) successfully works around the issue.

[1] https://github.com/osmandapp/OsmAnd-core/blob/2720aaebc98ebecfaf1a3f6b57fe97f2a0173133/src/GeoTiffCollection_P.cpp#L1040

### Steps to reproduce

Create an OsmAnd-compatible terrain geotiff of a rectangular area about 7000km^2. `Pixel Size = (7.735852665470224,-7.735852665470224)` was the lowest resolution tested, though it didn't seem that a different resolution had any impact.

### Actual result

The 3D terrain is only visible when zoomed out a certain amount. Zooming in a bit shows the same LOD but now blurry, zooming in even more doesn't show the 3D terrain at all. This is consistent with visibility / lack thereof of the grayscale representation of the geotiff (numeric `9` in `bird).

### Expected result

The 3D terrain continues to be visible when I zoom in.

### Your Environment (required)

I used `bird` with the latest commit of `osmand-core`.

**Comment by mariush444** [comment]
How to create own OsmAnd-compatible geotiff ?

**Comment by vshcherb** [comment]
It needs to be documented somewhere already or answered. I think only 1 criteria: OsmAnd compatible geotiff should be in Merkator projection

**Comment by Mis012** [comment]
@vshcherb: So, are you able to reproduce it? If it helps I could upload an example geotiff somewhere.

Also, it seems that I'm not able to get the same resolution / performance as in QGIS (on the same machine). Is there any chance you could look at the QGIS renderer to see what they are doing differently? Looking at OsmAnd, I'd almost think it's not worth supporting full resolution for 2m DEMs, but in QGIS on the same machine they work just fine and they look *really* nice, I think they could even make a real difference on a hike because the resolution is just *that* good.

**Comment by Andrewschenko** [comment]
Minor bug was fixed. Certainly, heightmaps shouldn't be displayed for particular zoom levels only (I guess, it's ninth level). [Here](https://github.com/osmandapp/OsmAnd/discussions/20543#discussioncomment-10325349) I've explained the conditions that must be met to show the tile. However, there is a small overlap (~ 5% for 3D relief and ~0.8% for rasters - hillshade, slope, etc.). This overlap is applied at each tile at its corners, that should also be taken into consideration every time one produce a heightmap tiff, so it needs to be made slightly larger in each of its 9-level tile.

**Comment by Mis012** [comment]
just for completeness, the full resolution I care about is `Pixel Size = (2.100320322851903,-2.100320322851903)`, it's just that for the downscaled version with ~7,-7 resolution I knew for sure the only difference should be the size of the area.

**Comment by Mis012** [comment]
@Andrewschenko: hm, the behavior seems less erratic now, but still problematic for the "small" geotiff. Up to a certain zoom level it doesn't disappear, but it has incorrect LOD, and after a certain zoom level it still disappears. The workaround doesn't have this issue. This is an example geotiff that can reproduce the issue: https://gofile.io/d/PnBQva

**Comment by Andrewschenko** [comment]
> @Andrewschenko: hm, the behavior seems less erratic now, but still problematic for the "small" geotiff. Up to a certain zoom level it doesn't disappear, but it has incorrect LOD, and after a certain zoom level it still disappears. The workaround doesn't have this issue. This is an example geotiff that can reproduce the issue: https://gofile.io/d/PnBQva

Such small geotiff as provided, which is not a set of 9-level tiles, unfortunately, can be shown as hillshade/slope rasters in lower (<9) zoom levels. In this case we simply don't check all the 9-level tiles included for the sake of performance.

**Comment by Mis012** [comment]
Is there a gdal command that can be used to work around this issue? (at the expense of making the file larger I imagine, though zero elevation tiles should compress nicely)

I specifically care about being able to "zoom in", ideally being able to utilize the full resolution of the dataset.

**Comment by Andrewschenko** [comment]
> Is there a gdal command that can be used to work around this issue? (at the expense of making the file larger I imagine, though zero elevation tiles should compress nicely)
> 
> I specifically care about being able to "zoom in", ideally being able to utilize the full resolution of the dataset.

You can use tiler.py from /tools this way:

tiler.py --size=35 --overlap=3 --zoom=9 --driver=GTiff
--driver-options="COMPRESS=LZW;PREDICTOR=3;SPARSE_OK=TRUE;TILED=YES;BLOCKXSIZE=80;BLOCKYSIZE=80"
--extension=tif --verbose "whole.tif" "/output_path"

Then combine them into single .vrt with gdalbuildvrt and translate this .vrt into result geotiff using gdal_translate

**Comment by Mis012** [comment]
pretty sure `--zoom=9` will downscale to the resolution used by OsmAnd, which kills the whole point of using a custom geotiff...

**Comment by Andrewschenko** [comment]
> pretty sure `--zoom=9` will downscale to the resolution used by the geotiffs that can already be downloaded through OsmAnd, which kills the whole point of using a custom geotiff... and iirc there are more places where downscaling is hardcoded.
> 
> I already have a working ~7,-7 resolution geotiff, afaict the only difference is the size, if there's a minimum geographic area then I guess knowing that would be enough. The full resolution ~2,-2 geotiff is 1.6GiB, so it's clearly not the size of the geotiff that's too small, but the area of the globe it represents.

No, tyler.py has nothing to do with the actual tiff resolution - it just slices your source geotiff into proper tiles of ninth zoom (with proper overlaps). You can easily use source tiff of any resolution and any size (even smaller, than zoom9), just make boundaries larger (use gdalbuildvrt for that) to slice it (vrt) then into proper tiles. Don't bother about size - empty areas cost nothing in this sense. The only restriction here (yet) is that OsmAnd calculates maximum zoom level of all provided geotiff files by the one with the smallest resolution.

**Comment by Mis012** [comment]
`tiler.py: error: Input file must have single raster band with Int16 type` pretty sure this would already limit the resolution...

**Comment by Andrewschenko** [comment]
> ~`tiler.py: error: Input file must have single raster band with Int16 type` pretty sure this would already limit the resolution...~ oh nvm, old version?
> 
> but it does say `Tiles to process: 0`, so I guess need to make the boundaries larger... but by how much? do I just add a degree on each side to be safe?

Yes, that would be enough. Zoom9 = 512x512 tiles for the world map.

**Comment by DmitryAlexei** [comment]
Needs to be documented

**Comment by vshcherb** [comment]
For documentation: 
- Create article for .geotif format which supported as DEM data in Application
- Describe how to convert from .geotif (with degree projection - input as we use from DEM) produce Merkator projection .geotif (supported in OsmAnd)

**Comment by Mis012** [comment]
>The only restriction here (yet) is that OsmAnd calculates maximum zoom level for all provided geotiff files by the one with the smallest resolution. Probably, this behaviour should be changed.

just remembered about this, would be nice if that got fixed
